### DISCUSSION TITLE
Engine magnetos

### DISCUSSION THREAD

**Discussion by bomber64** [JSBSim-Team/jsbsim#1112] [Support requests]
engine magnetos according to the manual is the property

propulsion/magneto_cmd

In flightgear it's not exposed in the property tree.

I have a 4 engines heavy bomber where I need to test each engines left right and both magneto functionality

Now I mentioned flightgear above but it's not for that simulation it's for another.

Is there anyway in the future I could have access to individual engines magnetos

regards

Simon

**Comment by seanmcleod** [comment]
Taking a look at the source code there is the ability to set the active engine.

https://github.com/JSBSim-Team/jsbsim/blob/c867cc7116283f08850db36444baf2577b967b67/src/models/FGPropulsion.cpp#L817-L818

Which is then used for magnetos, starter etc. Although there is no modelling of multiple magnetos per engine.

https://github.com/JSBSim-Team/jsbsim/blob/c867cc7116283f08850db36444baf2577b967b67/src/models/FGPropulsion.cpp#L592-L607

**Reply by bomber64** [reply]
Hi Sean, I'm not familiar with how that active engine code works as it's non functional in fightgear

Within the engine file exists this

> sparkfaildrop 0.1 /sparkfaildrop
>  this is the percentage drop in horsepower for single magneto operation, 10% equaals 0.1

within the propulsion properties exists this property...... active_engine (as you identified in the code)

If I set its value to 1, can I assume it's working on engine 1 only ?
How does magneto_cmd property then work ?
If I set it to 3 does that put both magentos on...

If I then change active_engine to 2 whilst magneto_cmd is at 3 would that then turn engine 2 magneto both on. ?

Sorry to be a pain, but struggling to visualise this and flightgears property tree doesn't help in this particular case

Simon

**Comment by seanmcleod** [comment]
So given the following property binding:

https://github.com/JSBSim-Team/jsbsim/blob/c867cc7116283f08850db36444baf2577b967b67/src/models/FGPropulsion.cpp#L814

That means anytime the property `propulsion/magneto_cmd` is written to with an integer value, whether from a JSBSim XML based script, via the Python interface etc. the method `FGPropulsion::SetMagnetos(int)` is called.

And as shown in the code snippet I posted in the earlier post you'll see that if the `ActiveEngine < 0` then the value written to the property is forwarded to all engines. Otherwise it's only forwarded to the active engine.

And by default:

https://github.com/JSBSim-Team/jsbsim/blob/c867cc7116283f08850db36444baf2577b967b67/src/models/FGPropulsion.cpp#L77

Ignore the following statement I made, I assumed `SetMagnetos(int)` was really a boolean, 0 or 1.

> Although there is no modelling of multiple magnetos per engine.

Whereas it's really:

https://github.com/JSBSim-Team/jsbsim/blob/c867cc7116283f08850db36444baf2577b967b67/src/models/propulsion/FGPiston.cpp#L564-L568
